packages/grpc-js/src/client-interceptors.ts

```diff
@@ -0,0 +1,563 @@
+}
+
+/**
+ * An object with methods for intercepting and modifying outgoing call operations.
+ */
+export interface FullRequester {
+  start: MetadataRequester;
+  sendMessage: MessageRequester;
+  halfClose: CloseRequester;
+  cancel: CancelRequester;
+}
+
+export type Requester = Partial<FullRequester>;
+
+export class ListenerBuilder {
+  private metadata: MetadataListener | undefined = undefined;
+  private message: MessageListener | undefined = undefined;
+  private status: StatusListener | undefined = undefined;
+
+  withOnReceiveMetadata(onReceiveMetadata: MetadataListener): this {
+    this.metadata = onReceiveMetadata;
+    return this;
+  }
+
+  withOnReceiveMessage(onReceiveMessage: MessageListener): this {
+    this.message = onReceiveMessage;
+    return this;
+  }
+
+  withOnReceiveStatus(onReceiveStatus: StatusListener): this {
+    this.status = onReceiveStatus;
+    return this;
+  }
+
+  build(): Listener {
+    return {
+      onReceiveMetadata: this.metadata,
+      onReceiveMessage: this.message,
+      onReceiveStatus: this.status,
+    };
+  }
+}
+
+export class RequesterBuilder {
+  private start: MetadataRequester | undefined = undefined;
+  private message: MessageRequester | undefined = undefined;
+  private halfClose: CloseRequester | undefined = undefined;
+  private cancel: CancelRequester | undefined = undefined;
+
+  withStart(start: MetadataRequester): this {
+    this.start = start;
+    return this;
+  }
+
+  withSendMessage(sendMessage: MessageRequester): this {
+    this.message = sendMessage;
+    return this;
+  }
+
+  withHalfClose(halfClose: CloseRequester): this {
+    this.halfClose = halfClose;
+    return this;
+  }
+
+  withCancel(cancel: CancelRequester): this {
+    this.cancel = cancel;
+    return this;
+  }
+
+  build(): Requester {
+    return {
+      start: this.start,
+      sendMessage: this.message,
+      halfClose: this.halfClose,
+      cancel: this.cancel,
+    };
+  }
+}
+
+/**
+ * A Listener with a default pass-through implementation of each method. Used
+ * for filling out Listeners with some methods omitted.
+ */
+const defaultListener: FullListener = {
+  onReceiveMetadata: (metadata, next) => {
+    next(metadata);
+  },
+  onReceiveMessage: (message, next) => {
+    next(message);
+  },
+  onReceiveStatus: (status, next) => {
+    next(status);
+  },
+};
+
+/**
+ * A Requester with a default pass-through implementation of each method. Used
+ * for filling out Requesters with some methods omitted.
+ */
+const defaultRequester: FullRequester = {
+  start: (metadata, listener, next) => {
+    next(metadata, listener);
+  },
+  sendMessage: (message, next) => {
+    next(message);
+  },
+  halfClose: next => {
+    next();
+  },
+  cancel: next => {
+    next();
+  },
+};
+
+export interface InterceptorOptions extends CallOptions {
+  // tslint:disable-next-line no-any
+  method_definition: ClientMethodDefinition<any, any>;
+}
+
+export interface InterceptingCallInterface {
```

Comment:
I considered that, but there are a couple of important differences in the APIs, specifically in the `start` and `sendMessageWithContext` methods, plus some methods that are only in the `CallInterface`. The `InterceptingCallInterface` is a public interface based on the original implementation, so it needs to work the same as that. The `CallInterface` is private, so it can have more rigid types, and it can have some extra methods that are intended for internal use, particularly by the `Channel` and `Subchannel` classes.

Reply:
That makes sense, thanks for the clarification!